apps/aeutils/src/aeu_mtrees.erl

```diff
@@ -0,0 +1,109 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc In-memory Merkle trees.
+%%%
+%%% The root hash depends on the order of the operations performed on
+%%% the tree.
+%%%
+%%% This module is a wrapper for `gb_merkle_trees`, implementing the
+%%% following enhancements:
+%%% * Stricter checks on arguments;
+%%% * Some less ambiguous return values;
+%%% * API compatible with the OTP `gb_trees` module;
+%%% * It enables better type specifications in code using this module.
+%%%
+%%% @see gb_merkle_trees
+%%% @see gb_trees
+%%% @end
+%%%-------------------------------------------------------------------
+-module(aeu_mtrees).
+
+%% API - subset of OTP `gb_trees` module
+-export([empty/0,
+         lookup/2,
+         enter/3,
+         to_list/1]).
+
+%% API - Merkle tree
+-export([root_hash/1,
+         lookup_with_proof/2,
+         verify_proof/4]).
+
+-export_type([mtree/0,
+              mtree/2,
+              root_hash/0,
+              proof/0]).
+
+-define(HASH_BYTES, 32).
+-define(IS_KEY(K), is_binary(K)).
+-define(IS_VALUE(V), is_binary(V)).
+
+-type key() :: binary().
+-type value() :: binary().
+-type mtree() :: mtree(key(), value()).
+
+%% Enable specification of types of key and value for enabling code
+%% using this module to document types for readability.
+%% Both key and value must be binaries.
+-type mtree(_K, _V) :: gb_merkle_trees:tree().
+
+-type root_hash() :: <<_:(?HASH_BYTES*8)>>.
+
+-type proof() :: gb_merkle_trees:merkle_proof().
+
+%%%===================================================================
+%%% API - subset of OTP `gb_trees` module
+%%%===================================================================
+
```

Comment:
I would like the get/2 as well. That crashes if the key is not present.

Reply:
Ok, I plan to update this PR with this function.

Reply:
Addressed in commit 6ccbeaeb added (no force-push, just added). Hence I would need your re-review @gorillainduction please.

Reply:
LGTM!